istioctl/pkg/writer/compare/sds/writer.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sdscompare
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+	"text/tabwriter"
+)
+
+// SDSWriter takes lists of SecretItem or SecretItemDiff and prints them through supplied output writer
+type SDSWriter interface {
+	PrintSecretItems([]SecretItem) error
+	PrintDiffs([]SecretItemDiff) error
+}
+
+type Format int
+
+const (
+	JSON Format = iota
+	TABULAR
+)
+
+// NewSDSWriter generates a new instance which conforms to SDSWriter interface
+func NewSDSWriter(w io.Writer, format Format) SDSWriter {
+	return &sdsWriter{
+		w:      w,
+		output: format,
+	}
+}
+
+// sdsWriter is provided concrete implementation of SDSWriter
+type sdsWriter struct {
+	w      io.Writer
+	output Format
+}
+
+// PrintSecretItems uses the user supplied output format to determine how to display the diffed secrets
+func (w *sdsWriter) PrintSecretItems(secrets []SecretItem) error {
+	var err error
+	switch w.output {
+	case JSON:
+		err = w.printSecretItemsJSON(secrets)
+	case TABULAR:
+		err = w.printSecretItemsTabular(secrets)
+	}
+	return err
+}
+
+var (
+	secretItemColumns = []string{"RESOURCE NAME", "TYPE", "VALID CERT", "SERIAL NUMBER", "EXPIRES", "NOT BEFORE", "SOURCE", "DESTINATION"}
```

Comment:
So the EXPIRES really is the `NOT AFTER` from x509 cert?

Reply:
Yes, I could also change the naming so that it matches exactly with the x509 field.

Reply:
That will be better.

Reply:
fixed :+1: